test: Fix inconsistent lock order in wallet_tests/CreateWallet

hebasto opened the pull request:
This PR:
- fixes #19049 that was caused by #16426
- removes `wallet_tests::CreateWallet` suppression from the `test/sanitizer_suppressions/tsan`

The example of the improved `CheckLastCritical()`/`LEAVE_CRITICAL_SECTION()` log (could be got when compiled without the last commit):
```
2020-09-20T08:34:28.429485Z [test] INCONSISTENT LOCK ORDER DETECTED
2020-09-20T08:34:28.429493Z [test] Current lock order (least recent first) is:
2020-09-20T08:34:28.429501Z [test]  'walletInstance->cs_wallet' in wallet/wallet.cpp:4007 (in thread 'test')
2020-09-20T08:34:28.429508Z [test]  'cs_wallets' in wallet/wallet.cpp:4089 (in thread 'test')
```

Currently, there are other "naked" `LEAVE_CRITICAL_SECTION()` in the code base:
https://github.com/bitcoin/bitcoin/blob/b99a1633b270e0e89479b2bb2ae19a8a8dc0fa05/src/rpc/mining.cpp#L698

https://github.com/bitcoin/bitcoin/blob/b99a1633b270e0e89479b2bb2ae19a8a8dc0fa05/src/checkqueue.h#L208

hebasto committed:
sync: Improve CheckLastCritical()

This commit adds actual lock stack logging if check fails.

hebasto committed:
[skip ci] sync: Check precondition in LEAVE_CRITICAL_SECTION() macro

This change reveals a bug in the wallet_tests/CreateWalletFromFile test,
that will be fixed in the following commit.

hebasto committed:
test: Fix inconsistent lock order in wallet_tests/CreateWallet

promag commented:
> This change reveals a bug in the wallet_tests/CreateWalletFromFile test,
that will be fixed in the following commit.

@hebasto I'm not following, how is it revealed?

hebasto commented:
@promag 
> > This change reveals a bug in the wallet_tests/CreateWalletFromFile test,
> > that will be fixed in the following commit.
> 
> @hebasto I'm not following, how is it revealed?

```
$ git checkout 7ac14ec58740226af85401f879f37f09755b990b
$ ./autogen.sh && ./configure --enable-debug && make clean && make
$ src/test/test_bitcoin --run_test=wallet_tests/CreateWalletFromFile
Running 1 test case...
wallet/test/wallet_tests.cpp:786 wallet->wallet()->cs_wallet was not most recent critical section locked, details in debug log.
unknown location(0): fatal error: in "wallet_tests/CreateWalletFromFile": signal: SIGABRT (application abort requested)
wallet/test/wallet_tests.cpp(785): last checkpoint

*** 1 failure is detected in the test module "Bitcoin Core Test Suite"
^C
$ tail -4 '/tmp/test_common_Bitcoin Core/03a41ec8724b4ea8adc603b0dc139290a6f9c765607df0bc277823d7b0165c20/regtest/debug.log'
2020-09-21T09:16:36.739551Z [test] INCONSISTENT LOCK ORDER DETECTED
2020-09-21T09:16:36.739557Z [test] Current lock order (least recent first) is:
2020-09-21T09:16:36.739562Z [test]  'walletInstance->cs_wallet' in wallet/wallet.cpp:4007 (in thread 'test')
2020-09-21T09:16:36.739568Z [test]  'cs_wallets' in wallet/wallet.cpp:4089 (in thread 'test')
```

promag commented:
Thanks @hebasto, I tried `make check` instead. I'd suggest to add `[skip ci]` to that commit body

hebasto commented:
@promag 
> Thanks @hebasto, I tried `make check` instead.

```
$ make check
...
2020-09-21T09:41:56.850457Z [test] INCONSISTENT LOCK ORDER DETECTED
2020-09-21T09:41:56.850463Z [test] Current lock order (least recent first) is:
2020-09-21T09:41:56.850470Z [test]  'walletInstance->cs_wallet' in wallet/wallet.cpp:4007 (in thread 'test')
2020-09-21T09:41:56.850475Z [test]  'cs_wallets' in wallet/wallet.cpp:4089 (in thread 'test')
wallet/test/wallet_tests.cpp:786 wallet->wallet()->cs_wallet was not most recent critical section locked, details in debug log.
make[3]: *** [Makefile:20887: wallet/test/wallet_tests.cpp.test] Error 1
make[3]: Leaving directory '/home/hebasto/guix/GitHub/bitcoin/src'
make[2]: *** [Makefile:19327: check-am] Error 2
make[2]: Leaving directory '/home/hebasto/guix/GitHub/bitcoin/src'
make[1]: *** [Makefile:19012: check-recursive] Error 1
make[1]: Leaving directory '/home/hebasto/guix/GitHub/bitcoin/src'
make: *** [Makefile:798: check-recursive] Error 1
```

hebasto commented:
Updated 283ddf104573fcf2ae648147e5ecf24c776246df -> d89c332c597a1129aa9fd3293bd37e7ab90171eb ([pr19982.03](https://github.com/hebasto/bitcoin/commits/pr19982.03) -> [pr19982.04](https://github.com/hebasto/bitcoin/commits/pr19982.04), [diff](https://github.com/hebasto/bitcoin/compare/pr19982.03..pr19982.04)):

- addressed @promag's [comment](https://github.com/bitcoin/bitcoin/pull/19982#issuecomment-696001691):
> I'd suggest to add `[skip ci]` to that commit body

promag commented:
Oh I did't have `--enable-debug`.

promag commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -776,16 +778,18 @@ BOOST_FIXTURE_TEST_CASE(CreateWalletFromFile, TestChain100Setup)
             m_coinbase_txns.push_back(CreateAndProcessBlock({block_tx}, GetScriptForRawPubKey(coinbaseKey.GetPubKey())).vtx[0]);
             mempool_tx = TestSimpleSpend(*m_coinbase_txns[3], 0, coinbaseKey, GetScriptForRawPubKey(key.GetPubKey()));
             BOOST_CHECK(chain->broadcastTransaction(MakeTransactionRef(mempool_tx), DEFAULT_TRANSACTION_MAXFEE, false, error));
+            LEAVE_CRITICAL_SECTION(cs_wallets);
```
Can't we refactor the test to remove these (LEAVE/ENTER)CRITICAL_SECTION? Looks like this block could go after `TestLoadWallet`?

hebasto replied:
I don't think so as `HandleLoadWallet()` puts a lambda closure into the `g_load_wallet_fns` which in turn is used by `CWallet::Create()`: https://github.com/bitcoin/bitcoin/blob/b99a1633b270e0e89479b2bb2ae19a8a8dc0fa05/src/wallet/wallet.cpp#L4090-L4092

ariard reviewed (commented):
Code Review ACK d89c332, feel free to take comments,  thanks for fixing this.

AFAICT this deadlock wasn't catch by CI at #16426 merge because LEAVE_CRITICAL didn't call neither `potential_deadlock_detected` or `CheckLastCritical`. Otherwise if a new lock would have been taken in the rest of `CreateWalletFromFile` it should have been detected.

So this PR does improve deadlock detection by flagging any lock order inconsistency even if it didn't materialize in an effective one as this case sounds to suggest ?

hebasto commented on src/sync.cpp:
```diff
@@ -182,20 +182,28 @@ void EnterCritical(const char* pszName, const char* pszFile, int nLine, void* cs
+        tfm::format(std::cerr, "%s:%s %s was not most recent critical section locked, details in debug log.\n", file, line, guardname);
+        abort();
+    }
+    throw std::logic_error(strprintf("%s was not most recent critical section locked", guardname));
```
Also it can throw in "naked" `LEAVE_CRITICAL_SECTION` that we have in the code base.

hebasto commented on src/test/sync_tests.cpp:
```diff
@@ -55,4 +75,26 @@ BOOST_AUTO_TEST_CASE(potential_deadlock_detected)
+
+    Mutex mutex1, mutex2;
+    TestInconsistentLockOrderDetected(mutex1, mutex2);
+    // The second test ensures that lock tracking data have not been broken by exception.
```
I'll address this comment if the reviewers request any other changes.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19982#issuecomment-727253125).

ryanofsky reviewed (approved):
Code review ACK d89c332c597a1129aa9fd3293bd37e7ab90171eb. Impressive work by hebasto not only debugging and fixing the buggy wallet test, but also improving deadlock detection, and improving deadlock detection output, and writing new deadlock detection tests!

@hebasto if you feel like adding more information to #19049, I'd be curious to know more you debugged this, since I gave up there struggling to figure out which locks were involved

ryanofsky commented:
Note to anyone else interested in reviewing this that the change is simpler than it probably looks.

The actual bugfix in the third commit is very straightforward and just two lines. The first two commits are just improvements to error handling and debug output in lock order code, ignoring a big whitespace diff in the first commit.

hebasto commented:
@ryanofsky 
> @hebasto if you feel like adding more information to #19049, I'd be curious to know more you debugged this, since I gave up there struggling to figure out which locks were involved

Our own lock stack being printed at some points appears very useful for debugging.

hebasto commented:
Rebased d89c332c597a1129aa9fd3293bd37e7ab90171eb -> e2dc75f88fed8a4c0ca4baab485715001dce03d0 ([pr19982.04](https://github.com/hebasto/bitcoin/commits/pr19982.04) -> [pr19982.05](https://github.com/hebasto/bitcoin/commits/pr19982.05)) due to the conflict with #20285.

Also updated comments as [requested](https://github.com/bitcoin/bitcoin/pull/19982#discussion_r493716292) by @ariard.

hebasto commented:
Rebased d15646b4accabc2129fb4fa7049fb09c8def1c56 -> f2933e7a8636df9d428f89afb7208d35bb4c4041 ([pr19982.06](https://github.com/hebasto/bitcoin/commits/pr19982.06) -> [pr19982.07](https://github.com/hebasto/bitcoin/commits/pr19982.07)) due to the conflict with #19337.

ryanofsky reviewed (approved):
Code review ACK f2933e7a8636df9d428f89afb7208d35bb4c4041. Since last review just rebase updating test name and tweaking comments

ryanofsky commented:
Notes that might be helpful to other reviewers: https://github.com/bitcoin/bitcoin/pull/19982#issuecomment-717520816.

The actual fix is very simple and contained in the third commit. The other two commits are error checking improvements.

vasild commented on src/test/reverselock_tests.cpp:
```diff
@@ -48,12 +48,22 @@ BOOST_AUTO_TEST_CASE(reverselock_errors)
+
     // Make sure trying to reverse lock a previous lock fails
+    bool error_thrown = false;
     try {
         REVERSE_LOCK(lock2);
         BOOST_CHECK(false); // REVERSE_LOCK(lock2) succeeded
-    } catch(...) { }
+    } catch(const std::logic_error& e) {
+        BOOST_CHECK_EQUAL(e.what(), "lock2 was not most recent critical section locked");
+        error_thrown = true;
+    }
+    BOOST_CHECK(error_thrown);
```
`BOOST_CHECK_EXCEPTION()` can help simplify this:

```suggestion
    BOOST_CHECK_EXCEPTION(REVERSE_LOCK(lock2), std::logic_error, HasReason("lock2 was not most recent critical section locked"));
```

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19982#issuecomment-737757832).

vasild commented on src/test/sync_tests.cpp:
```diff
@@ -64,6 +64,26 @@ void TestDoubleLock(bool should_throw)
+    ENTER_CRITICAL_SECTION(mutex1);
+    ENTER_CRITICAL_SECTION(mutex2);
+#ifdef DEBUG_LOCKORDER
+    bool error_thrown = false;
+    try {
+        LEAVE_CRITICAL_SECTION(mutex1);
+    } catch (const std::logic_error& e) {
+        BOOST_CHECK_EQUAL(e.what(), "mutex1 was not most recent critical section locked");
+        error_thrown = true;
+    }
+    BOOST_CHECK(error_thrown);
```
ditto:
```suggestion
    BOOST_CHECK_EXCEPTION(LEAVE_CRITICAL_SECTION(mutex1), std::logic_error, HasReason("mutex1 was not most recent critical section locked"));
```

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19982#issuecomment-737757832).

vasild reviewed (approved):
ACK f2933e7a

```
lock m1
lock m2
unlock m1
unlock m2
```

the above was allowed before this PR and is forbidden after the PR. Such code cannot cause deadlock or any other issue and is legit. However it looks strange, may be a signal of a problem elsewhere and better be avoided. In our code in particular it would also brick the debugging mechanisms because `LeaveCritical()` from `unlock m1` would remove the wrong lock from the lock stack (`m2`).

I think it is ok to enforce "locks should be released in reverse order in which they were acquired"

hebasto commented:
Updated f2933e7a8636df9d428f89afb7208d35bb4c4041 -> 608e14a2b69665b33df046cb96c74fefa2613969 ([pr19982.07](https://github.com/hebasto/bitcoin/commits/pr19982.07) -> [pr19982.08](https://github.com/hebasto/bitcoin/commits/pr19982.08), [diff](https://github.com/hebasto/bitcoin/compare/pr19982.07..pr19982.08)):

- addressed @vasild's comments

practicalswift commented:
Concept ACK

Nice work!

promag reviewed (commented):
Tested ACK 608e14a2b69665b33df046cb96c74fefa2613969.

As expected `make check` fails on 79bfd2e27280eb0ad794b1f3389f4b00b38376b3 with `./configure --enable-debug`.

vasild commented:
ACK 608e14a

hebasto commented:
Rebased 608e14a2b69665b33df046cb96c74fefa2613969 -> e1e68b6305beb47ebf7ee48f14e12fdebdfea1ef ([pr19982.08](https://github.com/hebasto/bitcoin/commits/pr19982.08) -> [pr19982.09](https://github.com/hebasto/bitcoin/commits/pr19982.09)) due to the conflict with #19425.

MarcoFalke commented:
review ACK e1e68b6305beb47ebf7ee48f14e12fdebdfea1ef 💂

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK e1e68b6305beb47ebf7ee48f14e12fdebdfea1ef 💂
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjlrAv/QIaj8D90XXyl/5Sw/WeG+cGB2diICRF1Pbv4qvEvG7sr+ETsScHeJvK2
TVlXBWujZcYo50jpPMl9j8XcnLR9MfnT0GYTwOi+z+fcXmi4NULQRqx6JrNQS0Cb
cZQXp/2BNo1cMBEKmUmfnG221SzZ+rKuhiJXBte3JprpIhQyJTMvSAwW1yoVwwfJ
P4OQp0gvW+kOt1+Gy3DK7u9DLN848blddAvwqKaH276oemN92iClv7O9YUJAs4qS
gWLnyFGOMhUICJVOKYzneSAWBM+xRC/Ujx1pE1V18uv+F9lqjZIS95V2OLEHu8CQ
OBRHx1LqruQOHvpS/SC3317KCv68QJtCuiz254G5zxGFsMBAl9wGY2XZjBYdiazD
7GTUVSXzVwa474uVJRPVnuLvKLF0IgqDGrNQo2t/x1kW+8qJ23UpuKOhJ+1nOReG
6la8BU2ng9UKcvlwJ6PqFN/0A/GlLeNzfGPGxT1OzRtyu/4wrFeDvPxvLd5jXyJx
Ea6eO6Vm
=A3Z6
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `b9199fb93572c3373d95d9880b18d20bf1295e74f013c4bf476dda6306523525  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108b9199fb93572c3373d95d9880b18d20bf1295e74f013c4bf476dda6306523525f010da2c62a757542fd879817580b266e15c08fff01004c131de85c554ea50a2c75fad89468308f1045fd2750af008e97fda1fa5d603840083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0103dd78594f830c93c940147a5cacf971c08f020980fb04d33d342b30a0a84444844216be116040cfcc12e36ee162f169576b89a08f1045fd2750bf008cc46e9b6be844c5d0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0104af3ad5f958442e38d0d5d3ae232b4f708f120e64c8b7a5843431bbb5d9d2da849a214b28785e454c2dd8ca339cae25d139dfd08f1045fd2750af008c758d6593bebbde30083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010e0515a94d692d5e2981ec609d3c55f3208f1045fd2750bf00850ad4303c4fb8ac60083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

ryanofsky reviewed (approved):
Code review ACK e1e68b6305beb47ebf7ee48f14e12fdebdfea1ef. Just trivial rebase and suggested switch to BOOST_CHECK_EXCEPTION since last review

vasild reviewed (approved):
ACK e1e68b630
